### ISSUE TITLE
Uupdate api-platform/core 3.2.6 error

### ISSUE THREAD

**Issue by Napalm4577** [api-platform/core#5998]
**API Platform version(s) affected**: 3.2.6

**Description**  

I have a project on Symfony 6.3. I updated with composer, the library "api-platform/core" from 3.2.5 to 3.2.6 and I get error "'Warning: Undefined array key '$ref'" when I try to display the page `/api`

**How to reproduce**  

Update library from 3.2.5 to 3.2.6 with composer

**Possible Solution**  

**Additional Context**  

```json
{
  "@context": "\/api\/contexts\/Error",
  "@type": "hydra:Error",
  "hydra:title": "An error occurred",
  "hydra:description": "Warning: Undefined array key \u0022$ref\u0022",
  "trace": [
    {
      "namespace": "",
      "short_class": "",
      "class": "",
      "type": "",
      "function": "",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/JsonSchema\/SchemaFactory.php",
      "line": 227,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\JsonSchema",
      "short_class": "SchemaFactory",
      "class": "ApiPlatform\\JsonSchema\\SchemaFactory",
      "type": "-\u003E",
      "function": "buildPropertySchema",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/JsonSchema\/SchemaFactory.php",
      "line": 142,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\JsonSchema",
      "short_class": "SchemaFactory",
      "class": "ApiPlatform\\JsonSchema\\SchemaFactory",
      "type": "-\u003E",
      "function": "buildSchema",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/Hydra\/JsonSchema\/SchemaFactory.php",
      "line": 70,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\Hydra\\JsonSchema",
      "short_class": "SchemaFactory",
      "class": "ApiPlatform\\Hydra\\JsonSchema\\SchemaFactory",
      "type": "-\u003E",
      "function": "buildSchema",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/OpenApi\/Factory\/OpenApiFactory.php",
      "line": 353,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\OpenApi\\Factory",
      "short_class": "OpenApiFactory",
      "class": "ApiPlatform\\OpenApi\\Factory\\OpenApiFactory",
      "type": "-\u003E",
      "function": "collectPaths",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/OpenApi\/Factory\/OpenApiFactory.php",
      "line": 97,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\OpenApi\\Factory",
      "short_class": "OpenApiFactory",
      "class": "ApiPlatform\\OpenApi\\Factory\\OpenApiFactory",
      "type": "-\u003E",
      "function": "__invoke",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/Symfony\/Bundle\/SwaggerUi\/SwaggerUiProvider.php",
      "line": 75,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\Symfony\\Bundle\\SwaggerUi",
      "short_class": "SwaggerUiProvider",
      "class": "ApiPlatform\\Symfony\\Bundle\\SwaggerUi\\SwaggerUiProvider",
      "type": "-\u003E",
      "function": "provide",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/Symfony\/Security\/State\/AccessCheckerProvider.php",
      "line": 53,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\Symfony\\Security\\State",
      "short_class": "AccessCheckerProvider",
      "class": "ApiPlatform\\Symfony\\Security\\State\\AccessCheckerProvider",
      "type": "-\u003E",
      "function": "provide",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/State\/Provider\/DeserializeProvider.php",
      "line": 47,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\State\\Provider",
      "short_class": "DeserializeProvider",
      "class": "ApiPlatform\\State\\Provider\\DeserializeProvider",
      "type": "-\u003E",
      "function": "provide",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/Symfony\/Security\/State\/AccessCheckerProvider.php",
      "line": 53,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\Symfony\\Security\\State",
      "short_class": "AccessCheckerProvider",
      "class": "ApiPlatform\\Symfony\\Security\\State\\AccessCheckerProvider",
      "type": "-\u003E",
      "function": "provide",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/Symfony\/Validator\/State\/QueryParameterValidateProvider.php",
      "line": 42,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\Symfony\\Validator\\State",
      "short_class": "QueryParameterValidateProvider",
      "class": "ApiPlatform\\Symfony\\Validator\\State\\QueryParameterValidateProvider",
      "type": "-\u003E",
      "function": "provide",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/Symfony\/Validator\/State\/ValidateProvider.php",
      "line": 32,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\Symfony\\Validator\\State",
      "short_class": "ValidateProvider",
      "class": "ApiPlatform\\Symfony\\Validator\\State\\ValidateProvider",
      "type": "-\u003E",
      "function": "provide",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/Symfony\/Security\/State\/AccessCheckerProvider.php",
      "line": 53,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\Symfony\\Security\\State",
      "short_class": "AccessCheckerProvider",
      "class": "ApiPlatform\\Symfony\\Security\\State\\AccessCheckerProvider",
      "type": "-\u003E",
      "function": "provide",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/State\/Provider\/ContentNegotiationProvider.php",
      "line": 56,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\State\\Provider",
      "short_class": "ContentNegotiationProvider",
      "class": "ApiPlatform\\State\\Provider\\ContentNegotiationProvider",
      "type": "-\u003E",
      "function": "provide",
      "file": "\/var\/www\/isia-core\/vendor\/api-platform\/core\/src\/Documentation\/Action\/EntrypointAction.php",
      "line": 48,
      "args": []
    },
    {
      "namespace": "ApiPlatform\\Documentation\\Action",
      "short_class": "EntrypointAction",
      "class": "ApiPlatform\\Documentation\\Action\\EntrypointAction",
      "type": "-\u003E",
      "function": "__invoke",
      "file": "\/var\/www\/isia-core\/vendor\/symfony\/http-kernel\/HttpKernel.php",
      "line": 181,
      "args": []
    },
    {
      "namespace": "Symfony\\Component\\HttpKernel",
      "short_class": "HttpKernel",
      "class": "Symfony\\Component\\HttpKernel\\HttpKernel",
      "type": "-\u003E",
      "function": "handleRaw",
      "file": "\/var\/www\/isia-core\/vendor\/symfony\/http-kernel\/HttpKernel.php",
      "line": 76,
      "args": []
    },
    {
      "namespace": "Symfony\\Component\\HttpKernel",
      "short_class": "HttpKernel",
      "class": "Symfony\\Component\\HttpKernel\\HttpKernel",
      "type": "-\u003E",
      "function": "handle",
      "file": "\/var\/www\/isia-core\/vendor\/symfony\/http-kernel\/Kernel.php",
      "line": 197,
      "args": []
    },
    {
      "namespace": "Symfony\\Component\\HttpKernel",
      "short_class": "Kernel",
      "class": "Symfony\\Component\\HttpKernel\\Kernel",
      "type": "-\u003E",
      "function": "handle",
      "file": "\/var\/www\/isia-core\/public\/app_dev.php",
      "line": 28,
      "args": []
    }
  ]
}
```

Thanks guys

**Comment by soyuka** [comment]
I'd like to reproduce this, can you share your resources?

**Comment by Napalm4577** [comment]
Hello @soyuka,

What do you mean by "Resources" ?

**Comment by soyuka** [comment]
the classes you have `ApiResource` on that trigger this issue.

**Comment by ihmels** [comment]
@soyuka The error is caused by PR #5989. Previously, the method `SchemaFactory::buildSchema()` was always passed `Schema::TYPE_OUTPUT` as `$type` if it was called from the method `SchemaFactory::buildPropertySchema()`. Now the `$parentType`, which has the value `Schema::TYPE_INPUT` during schema generation, is always used. In `SchemaFactory`, line 86, this always leads to a return.

**Comment by ihmels** [comment]
@soyuka Your commit https://github.com/soyuka/core/commit/85e6209b6d944209b46ea207b8893ad41b0348d9 will fix this issue, but then there's still a bug:

Before 3.2.6:

```json
{
    "components": {
        "schemas": {
            "Product.SaveProduct-product.write": {
                "type": "object",
                "description": "",
                "deprecated": false,
                "properties": {
                    "codes": {
                        "type": "array",
                        "items": {
                            "$ref": "#\/components\/schemas\/ProductCode-product.write"
                        }
                    }
                }
            }
        }
    }
}
```

After your commit:

```json
{
    "components": {
        "schemas": {
            "Product.SaveProduct-product.write": {
                "type": "object",
                "description": "",
                "deprecated": false,
                "properties": {
                    "codes": {
                        "type": "array",
                        "items": {
                            "type": "unknown_type"
                        }
                    }
                }
            }
        }
    }
}
```

**Comment by soyuka** [comment]
Thanks this is exactly what I needed, I knew there would be some edge case with that fix, tyvm!

**Comment by ihmels** [comment]
You're welcome and thank you for this awesome API framework!

**Comment by Napalm4577** [comment]
Thanks @ihmels to give a basic example. Mine is very complex :smile:

**Comment by mbrodala** [comment]
I can confirm that the `$ref` error is gone with https://github.com/soyuka/core/commit/85e6209b6d944209b46ea207b8893ad41b0348d9

**Comment by soyuka** [comment]
@mbrodala it was also introducing an unkown-type I did a proper fix see https://github.com/api-platform/core/pull/6001 I would love some feedbacks if you can test it, thanks!

**Comment by Napalm4577** [comment]
Thanks @soyuka 

How we know when the fix will be released, maybe in the 3.2.7 ?

Thanks
